flax/linen/recurrent.py

```diff
@@ -497,3 +500,189 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
+        step, or replicated such that its values remain the same at each step. This
+        argument will be passed to `nn.scan`.
+
+    **Input shape**
+    
+    `x` should be an ndarray of shape `(b0...m, t, d0...dn)`where `b` are the batch
+    dimension, `t` is the time dimension, and `d` are the feature dimensions. The time
+    dimension can be in any position, `time_axis` should specify the corresponding
+    axis, internally the input will be reshaped to `(t, b0...m, d0...dn)`. The feature
+    dimensions should be last (after the time axis is to the front) and are specified by
+    `num_feature_dims`. The batch dimensions will simply be the remaining dimensions that
+    are not the time axis or the feature dimensions.
+
+    **Output shape**
+
+    If `return_sequences` is True, the output will have the same as the input, else
+    the time dimension will be dropped when the last time step is selected.
+    If `return_state` is True, the output will be a tuple of the output and the
+    carry.
+    """
+
+    if num_feature_dims < 1:
+      raise ValueError(f'num_feature_dims must be >= 1, got: {num_feature_dims}')
+    elif num_feature_dims >= len(x.shape):
+      raise ValueError(
+        f'\'num_feature_dims\' dimensions must be smaller than ' 
+        f'number of dimensions, got num_feature_dims \'{num_feature_dims}\' '
+        f'for input of shape \'{x.shape}\''
+      )
+
+    num_batch_dims = x.ndim - (num_feature_dims + 1) # features dims + time dim
+
+    # move time dimension to the first axis
+    x = x.swapaxes(0, time_axis)
+
+    # WARNING: this definition of initialization state
+    # is known to be problematic: https://github.com/google/flax/issues/652#issuecomment-1124216543
+    # Epecifically, this module will not update its cache if 
+    # 'params' are mutable, which can arise if users use `mutable=True`.
+    is_initializing = self.is_mutable_collection("params")
+
+    # get carry
+    if initial_state is not None:
+      initial_carry = initial_state
+    elif stateful and self.has_variable('cache', 'carry'):
+      initial_carry = self.get_variable('cache', 'carry')
+    else:
+      if num_feature_dims == 1:
+        size = x.shape[-1] # features dim
+      else:
+        size = x.shape[-num_feature_dims:] # spatial + feature dim
+      
+      initial_carry = cell.initialize_carry(
+        init_key,
+        batch_dims=x.shape[1 : 1 + num_batch_dims], 
+        size=size,
+      )
+
+    scan = transforms.scan(
+      cell.__class__.__call__,
+      in_axes=0, out_axes=0,
+      reverse=go_backwards,
+      # scan args
+      variable_axes=variable_axes,
+      variable_broadcast=variable_broadcast,
+      variable_carry=variable_carry,
+      split_rngs=split_rngs,
+    )
+
+    carry, x = scan(cell, initial_carry, x)
+
+    # if the module is initializing keep the initial carry
+    if is_initializing:
+      carry = initial_carry
+
+    if stateful:
+      self.put_variable('cache', 'carry', carry)
```

Comment:
Is this really cache or should the name be something more discriptive like "memory"?

Reply:
`memory` sounds good but I don't see a previous use of this collection, are we ok adding a new collection type?